### PR TITLE
refactor: update CDA configuration

### PR DESCRIPTION
**Pull request by saranyailla**
**Issue #, if available:**

**Description of changes:**
- Model  the CA config of CDA config and also include BYO CA keys. 
- Pass the CAConfiguration object to cert manager

**Why is this change necessary:**
- Needed to support BYOCA feature. 
- Refactoring code

**How was this change tested:**

**Any additional information or context required to review the change:**

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/test/resources/com/aws/greengrass/clientdevices/auth/config_with_ec_ca.yaml [hunk 1]
```diff
@@ -2,10 +2,10 @@
 services:
   aws.greengrass.clientdevices.Auth:
     configuration:
-      ca_type: ["ECDSA_P256"]
+      certificateAuthority:
+        ca_type: ["ECDSA_P256"]
   main:
     lifecycle:
-      install:
-        all: echo All installed
```
Comment (MikeDombo) [inline src/test/resources/com/aws/greengrass/clientdevices/auth/config_with_ec_ca.yaml:9]: why is this changed? this isn't a valid recipe anymore.

Comment (MikeDombo) [reply src/test/resources/com/aws/greengrass/clientdevices/auth/config_with_ec_ca.yaml:9]: Not that it needs to be a valid recipe, you can just remove the echo thing.

Comment (saranyailla) [reply src/test/resources/com/aws/greengrass/clientdevices/auth/config_with_ec_ca.yaml:9]: yeah, tests using invalid recipes have unnecessary component broken logs.

**Review by MikeDombo — DISMISSED** [review]

File: src/main/java/com/aws/greengrass/clientdevices/auth/ClientDevicesAuthService.java [hunk 2]
```diff
@@ -137,6 +140,7 @@ public ClientDevicesAuthService(Topics topics, GroupManager groupManager, Certif
         SessionCreator.registerSessionFactory("mqtt", mqttSessionFactory);
         certificateManager.updateCertificatesConfiguration(new CertificatesConfig(this.getConfig()));
         sessionManager.setSessionConfig(new SessionConfig(this.getConfig()));
+        certificateManager.setCertificateStoreConfig(this.getConfig(), this.getRuntimeConfig());
```
Comment (jbutler) [inline src/main/java/com/aws/greengrass/clientdevices/auth/ClientDevicesAuthService.java:143]: Take a look at how we deserialize GroupDefinition - ideally we can just follow a similar pattern here.

A couple benefits (I can think of):
* Don't need to plumb `Topics` all the way through. No need to deal with casting. We can add appropriate getters
* Config is immutable. Behavior is easier to reason about
* No need to add a bunch of static strings to this class and import them in other places

File: src/main/java/com/aws/greengrass/clientdevices/auth/certificate/CertificateStore.java [hunk 3]
```diff
@@ -83,6 +83,7 @@ public CertificateStore(Path workPath) {
         this.workPath = workPath;
     }
 
+
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/clientdevices/auth/certificate/CertificateStore.java:86]: nit: remove

File: src/main/java/com/aws/greengrass/clientdevices/auth/configuration/CAConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+public class CAConfiguration {
+    private final Topics certificateAuthorityTopics;
+    private final Topics cdaRuntimeTopics;
+
+    /**
+     * Construct CA Configuration object from the CDA component configuration.
+     *
+     * @param cdaConfigTopics CDA service configuration topics
+     */
+    @Inject
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/clientdevices/auth/configuration/CAConfiguration.java:32]: I don't think you can actually inject this.

File: src/main/java/com/aws/greengrass/clientdevices/auth/ClientDevicesAuthService.java [hunk 5]
```diff
@@ -63,6 +64,9 @@
 public class ClientDevicesAuthService extends PluginService {
     public static final String CLIENT_DEVICES_AUTH_SERVICE_NAME = "aws.greengrass.clientdevices.Auth";
     public static final String DEVICE_GROUPS_TOPICS = "deviceGroups";
+    public static final String CERTIFICATE_AUTHORITY_TOPIC = "certificateAuthority";
```
Comment (jbutler) [inline src/main/java/com/aws/greengrass/clientdevices/auth/ClientDevicesAuthService.java:67]: Move sub-keys of `certificateAuthority` into CAConfiguration

File: src/main/java/com/aws/greengrass/clientdevices/auth/configuration/CAConfiguration.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+import static com.aws.greengrass.clientdevices.auth.ClientDevicesAuthService.CA_CERTIFICATE_URI;
+import static com.aws.greengrass.clientdevices.auth.ClientDevicesAuthService.CA_PASSPHRASE;
+import static com.aws.greengrass.clientdevices.auth.ClientDevicesAuthService.CA_PRIVATE_KEY_URI;
+import static com.aws.greengrass.clientdevices.auth.ClientDevicesAuthService.CA_TYPE_TOPIC;
+import static com.aws.greengrass.clientdevices.auth.ClientDevicesAuthService.CERTIFICATE_AUTHORITY_TOPIC;
+import static com.aws.greengrass.componentmanager.KernelConfigResolver.CONFIGURATION_CONFIG_KEY;
+import static com.aws.greengrass.lifecyclemanager.GreengrassService.RUNTIME_STORE_NAMESPACE_TOPIC;
+
+public class CAConfiguration {
+    private final Topics certificateAuthorityTopics;
```
Comment (jbutler) [inline src/main/java/com/aws/greengrass/clientdevices/auth/configuration/CAConfiguration.java:23]: Now take this one step further and make this immutable. No need to store topics. We want `CertificateManager` and `CertificateStore` to always be operating on the same config. You can't achieve that if this class is dynamically reading from config

File: src/main/java/com/aws/greengrass/clientdevices/auth/certificate/CertificateStore.java (hunk 3 above)
Comment (jbutler) [reply src/main/java/com/aws/greengrass/clientdevices/auth/certificate/CertificateStore.java:86]: I'm also going to nitpick on this. Please remove

File: src/main/java/com/aws/greengrass/clientdevices/auth/CertificateManager.java [hunk 7]
```diff
@@ -70,6 +74,10 @@ public void updateCertificatesConfiguration(CertificatesConfig certificatesConfi
         this.certificatesConfig = certificatesConfig;
     }
 
+    public void setCAConfiguration(CAConfiguration caConfiguration) {
```
Comment (jbutler) [inline src/main/java/com/aws/greengrass/clientdevices/auth/CertificateManager.java:77]: Right now you have `setCAConfiguration` and `updateCAConfiguration` -- remove one of them

File: src/main/java/com/aws/greengrass/clientdevices/auth/configuration/CAConfiguration.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    private String caCertificateUri;
+    private List<String> caTypeList;
+    private String caPassphrase;
+
+    /**
+     * Updates the CA configuration with the latest CDA config.
+     *
+     * @param cdaConfigTopics CDA service configuration topics
+     */
+    public synchronized void setCAConfiguration(Topics cdaConfigTopics) {
```
Comment (jbutler) [inline src/main/java/com/aws/greengrass/clientdevices/auth/configuration/CAConfiguration.java:34]: Actually, this class still isn't immutable. Think about the cert rotation workflow - it's still possible for that to trigger during a config update which could result in us getting into a non-deterministic state.

I actually want you to re-instantiate this config class whenever there is a config change. Everything should be final

**Review by jbutler — APPROVED** [review]

File: src/main/java/com/aws/greengrass/clientdevices/auth/CertificateManager.java [hunk 9]
```diff
@@ -43,6 +44,9 @@ public class CertificateManager {
     private final Clock clock;
     private final Map<GetCertificateRequest, CertificateGenerator> certSubscriptions = new ConcurrentHashMap<>();
     private CertificatesConfig certificatesConfig;
+    @SuppressWarnings("PMD.UnusedPrivateField")
+    private CAConfiguration caConfiguration;
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/clientdevices/auth/CertificateManager.java:48]: future PR: read/write access to this needs to be synchronized or else you may be reading from multiple configuration instances.

**Review by MikeDombo — APPROVED** [review]
